### PR TITLE
Ca feature/self update

### PR DESCRIPTION
**Pull request by cagodoy**
Implementation for auto self updating

Features:
* Possibility of upgrade to mayor version
* Possibility for downgrade to minnor version.
* Custom encryptation. Now SHA256
* Uploader new operator: upload a specific operator version to S3 bucket. **This command overwrites the current bucket file. See attached image**
`./bin/uploader <version>`
* MQTT Implementation. `/update {version}`
* HTTP Implementation. `[POST] /update {version}`
* Makefile new commands:
    * clean-all: remove all files to bin folder
    * new-release: build pi operator and upload to S3 with checksum file
    * build-uploader: in charge to compile the uploader
    * checksum: in charge to generate a checksum file with respective version

On the other hand, now it has support for the route `/restart-service` (POST method in the case of HTTP) service through the HTTP and MQTT protocols.

----------------------

screen shot 2018-01-23 at 17 34 30

### CODE REVIEW

File: http.go [hunk 1]
```diff
@@ -44,13 +46,18 @@ type mqttStatus struct {
...
+}
+
 func newHealthResponse() *healthResponse {
 	data := new(healthzDataResponse)
 	data.Services = processManager.Services
 	data.Daemons = daemonStore
 
 	meta := new(healthzMetaResponse)
 	meta.MQTTStatus.IsConnected = processManager.MQTTClient.IsConnected()
+	meta.Version = version
```
Comment (mtavano) [inline http.go:60]: where is _version_ declared???

File: services.go [hunk 2]
```diff
@@ -293,6 +293,28 @@ func (ss *ServiceStore) StopService(name string) error {
...
+		return fmt.Errorf("services: service with name %q not found", name)
+	}
+
+	svc.logger().Info("Restarting")
+	if err := svc.Stop(); err != nil {
+		return err
+	}
+
+	cmd := svc.cmd
+	cmd = cmd.Clone()
```
Comment (mtavano) [inline services.go]: I think this two lines should be summarized as:

```Go
cmd := svc.cmd.Clone()
```

File: updater.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	var urlChecksum = up.baseURL + newVersion + ".checksum"
+	var urlBin = up.baseURL + newVersion
+
+	log := logger.GetLogger()
+	log.Info("New Version: ", newVersion)
+	log.Info("URL Checksum: ", urlChecksum)
+	log.Info("URL Bin: ", urlBin)
+
+	//define http client
+	client := http.Client{}
```
Comment (mtavano) [inline updater.go:47]: I think that client must be a pointer to an `http.Client` instead a simple instance. And, if your client has no custom configs of specific parameter, maybe  you should try with `client := http.DefaultClient`

File: updater.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	checksum = strings.TrimSpace(checksum)
+	respChecksum.Body.Close()
+
+	//get binary
+	log.Info("[Download Binary]")
+	respBin, err := client.Get(urlBin)
+	if err != nil {
+		return err
+	}
+	if respBin.StatusCode != 200 {
```
Comment (mtavano) [inline updater.go:75]: all http statuses different to a 200 means the same? are alls reffered to the checksum?

File: http.go (hunk 1 above)
Comment (cagodoy) [reply http.go:60]: @mtavano operator version, this is a meta

File: updater.go (hunk 3 above)
Comment (cagodoy) [reply updater.go:47]: lo intentamos pero no funcionó

File: updater.go (hunk 4 above)
Comment (cagodoy) [reply updater.go:75]: es para ver si existe solamente. con un 200 me basta.

File: updater.go (hunk 3 above)
Comment (mtavano) [reply updater.go:47]: read the compiler messages

**Pull request merged**